File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/MovingState.java
package net.cyclestreets.liveride;

import net.cyclestreets.CycleStreetsPreferences;
import net.cyclestreets.routing.Journey;

import org.osmdroid.util.GeoPoint;

abstract class MovingState extends LiveRideState
{
  private final int transition_;

  MovingState(final LiveRideState previous, final int transitionThreshold) {
    super(previous);
    transition_ = transitionThreshold;
  }

  @Override
  public final LiveRideState update(final Journey journey, final GeoPoint whereIam, final int accuracy) {
    int distanceFromEnd = journey.activeSegment().distanceFromEnd(whereIam);
    distanceFromEnd -= accuracy;
    if (distanceFromEnd < transition_)
      return transitionState(journey);

    return checkCourse(journey, whereIam, accuracy);
  }

  protected abstract LiveRideState transitionState(final Journey journey);

  private LiveRideState checkCourse(final Journey journey, final GeoPoint whereIam, final int accuracy) {
    int distance = journey.activeSegment().distanceFrom(whereIam);
    distance -= accuracy;

    if (distance > CycleStreetsPreferences.replanDistance())
      return new ReplanFromHere(this, whereIam);

    if (distance > CycleStreetsPreferences.offtrackDistance())
      return new GoingOffCourse(this);

    return this;
  }

  @Override
  public boolean isStopped() { return false; }
  @Override
  public boolean arePedalling() { return true; }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/MovingState.kt
package net.cyclestreets.liveride

import net.cyclestreets.CycleStreetsPreferences
import net.cyclestreets.routing.Journey
import org.osmdroid.util.GeoPoint

internal abstract class MovingState(previous: LiveRideState, private val transition_: Int) :
        LiveRideState(previous) {

    override fun update(journey: Journey, myLocation: GeoPoint, accuracy: Int): LiveRideState {
        var distanceFromEnd = journey.activeSegment()!!.distanceFromEnd(myLocation)
        distanceFromEnd -= accuracy

        if (distanceFromEnd < transition_) {
            return transitionState(journey)
        }
        return checkCourse(journey, myLocation, accuracy)
    }

    protected abstract fun transitionState(journey: Journey): LiveRideState

    private fun checkCourse(journey: Journey, myLocation: GeoPoint, accuracy: Int): LiveRideState {
        var distance = journey.activeSegment()!!.distanceFrom(myLocation)
        distance -= accuracy

        if (distance > CycleStreetsPreferences.replanDistance()) {
            return ReplanFromHere(this, myLocation)
        }
        if (distance > CycleStreetsPreferences.offtrackDistance()) {
            return GoingOffCourse(this)
        }
        return this
    }

    override fun isStopped(): Boolean { return false }
    override fun arePedalling(): Boolean { return true }
}
